# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaData.cpp
# Commit: 0d9c1ac74d9c
# Full Hash: 0d9c1ac74d9cf147aa99fad91c7e1eca0e24fa1b
# Author: Paul Adenot <paul@paul.cx>
# Date: 2024-02-02 04:43:27
# Description:
#   Bug 1874723 - Accept imprecision when trimming audio content media in certain cases. r=alwu
#   
#   In the test case, the time base is 69 (generated by a fuzzer), and so there's a
#   very large loss in precision in the computation, and eventually the assumption
#   doesn't hold. It should however hold for the common case.
# ==============================================================================

diff -r f84c76f9460d -r 0d9c1ac74d9c dom/media/MediaData.cpp
--- a/dom/media/MediaData.cpp	Thu Feb 01 17:10:37 2024 +0000
+++ b/dom/media/MediaData.cpp	Thu Feb 01 17:13:15 2024 +0000
@@ -107,9 +107,16 @@
   mDataOffset = frameOffset * mChannels;
   MOZ_DIAGNOSTIC_ASSERT(mDataOffset <= mAudioData.Length(),
                         "Data offset outside original buffer");
-  mFrames = (trimAfter - trimBefore).ToTicksAtRate(mRate);
-  MOZ_DIAGNOSTIC_ASSERT(mFrames <= mAudioData.Length() / mChannels,
-                        "More frames than found in container");
+  int64_t frameCountAfterTrim = (trimAfter - trimBefore).ToTicksAtRate(mRate);
+  if (frameCountAfterTrim >
+      AssertedCast<int64_t>(mAudioData.Length() / mChannels)) {
+    // Accept rounding error caused by an imprecise time_base in the container,
+    // that can cause a mismatch but not other kind of unexpected frame count.
+    MOZ_RELEASE_ASSERT(!trimBefore.IsBase(mRate));
+    mFrames = 0;
+  } else {
+    mFrames = frameCountAfterTrim;
+  }
   mTime = mOriginalTime + trimBefore;
   mDuration = TimeUnit(mFrames, mRate);
 
@@ -449,21 +456,21 @@
 
 nsCString VideoData::ToString() const {
   std::array ImageFormatStrings = {
-    "PLANAR_YCBCR",
-    "NV_IMAGE",
-    "SHARED_RGB",
-    "MOZ2D_SURFACE",
-    "MAC_IOSURFACE",
-    "SURFACE_TEXTURE",
-    "D3D9_RGB32_TEXTURE",
-    "OVERLAY_IMAGE",
-    "D3D11_SHARE_HANDLE_TEXTURE",
-    "D3D11_TEXTURE_IMF_SAMPLE",
-    "TEXTURE_WRAPPER",
-    "D3D11_YCBCR_IMAGE",
-    "GPU_VIDEO",
-    "DMABUF",
-    "DCOMP_SURFACE",
+      "PLANAR_YCBCR",
+      "NV_IMAGE",
+      "SHARED_RGB",
+      "MOZ2D_SURFACE",
+      "MAC_IOSURFACE",
+      "SURFACE_TEXTURE",
+      "D3D9_RGB32_TEXTURE",
+      "OVERLAY_IMAGE",
+      "D3D11_SHARE_HANDLE_TEXTURE",
+      "D3D11_TEXTURE_IMF_SAMPLE",
+      "TEXTURE_WRAPPER",
+      "D3D11_YCBCR_IMAGE",
+      "GPU_VIDEO",
+      "DMABUF",
+      "DCOMP_SURFACE",
   };
 
   nsCString rv;